### ISSUE TITLE
Drop impls for DOM types should be forbidden

### ISSUE THREAD

**Issue by jdm** [servo/servo#26488]
@pcwalton @Manishearth @asajeffrey I recall discussing this topic in Berlin, but I can't find the notes I thought I took. Attempting to reconstruct from memory:
* there exist DOM objects that have Drop impls for various reasons
* there was agreement that it should be possible to run some code in response to a GC finalizer (eg. webgl objects need to be able to unreference their graphics resources)
* the easiest way to ban Drop impls is to make the `#[dom_struct]` macro also add an empty Drop impl, which will conflict with any manual implementation
* for any code which does require a valid Drop, we can create inner structs that only hold the data required for that implementation and nothing more

Does this make sense? My memory is haziest when it comes to the last point.

**Comment by Manishearth** [comment]
This is correct. We add Drop impls, and we can use inner structs for everything else.

**Comment by Manishearth** [comment]
Worth noting: this isn't 100% sound given that inner structs that contain DOM types will still technically be unsafe to add destructors to.

**Comment by jdm** [comment]
What if we add an empty Drop impl for the `#[must_root]` attribute?

**Comment by Manishearth** [comment]
That works. Paging this back in, i have a vague recollection of us deciding to do that already.

**Comment by jdm** [comment]
Modifying the output of dom_struct is straightforward: https://github.com/servo/servo/blob/master/components/dom_struct/lib.rs
Modifying the output of #[must_root] is more complicated - I think we need to turn it into a macro like dom_struct, which outputs two items:
1. an empty Drop impl
1. a new `#[must_root::must_root_type]`, and the must_root pass in [this code](https://github.com/servo/servo/blob/4d541e8e381e3b3388dd1d1378d147e22197b11d/components/script_plugins/lib.rs#L116) would need to check for the new attribute instead

**Comment by splav** [comment]
@highfive: assign me

**Comment by splav** [comment]
Actually there are 27 current Drop implementation for various Dom structs. What should I do about them?
https://gist.githubusercontent.com/splav/a0dc65c5f395bc94d24294733d6aed30/raw/8d9477f674fa2773ef2e3f5f7462ab36636c7c15/gistfile1.txt

**Comment by jdm** [comment]
To allow the build to succeed, for each of those we can create a struct that only contains the members that are accessed in the Drop implementation, and move the Drop to be implemented on that new struct instead, and store an instance of that struct in the DOM object. It will then be easier to reason about whether each implementation is safe or not.

**Comment by nicoabie** [comment]
@highfive: assign me

**Comment by nicoabie** [comment]
@jdm we could create another macro to add
```
#[derive(JSTraceable, MallocSizeOf)]
#[unrooted_must_root_lint::must_root]
```
to other *DroppableFields, what do you think?

**Comment by keiya01** [comment]
@jdm 
May I try this issue?
This is the first time to contribute to Servo for me.

**Comment by keiya01** [comment]
@highfive: assign me

**Comment by keiya01** [comment]
Why do we need `#[must_root::must_root_type]` ?

**Comment by splav** [comment]
This is a new marker attribute to make the lint check work.

**Comment by keiya01** [comment]
What is the feature of `#[must_root::must_root_type]` ?

**Comment by keiya01** [comment]
> Modifying the output of #[must_root] is more complicated - I think we need to turn it into a macro like dom_struct, which outputs two items:
>
> 1. an empty Drop impl
> 2. a new #[must_root::must_root_type], and the must_root pass in this code would need to check for the new attribute instead

I think `#[must_root]` is a lint function but  `Drop impl` should be outputted by `#[must_root]`?

**Comment by sagudev** [comment]
I think to make it fully safe we need to forbid any js managed field in inner struct.

**Comment by willypuzzle** [comment]
@jdm could I work on this?

**Comment by jdm** [comment]
@willypuzzle Yes! The next step here is rebasing the work from https://github.com/servo/servo/pull/28302, and it would be good to work on it incrementally, rather than one large PR. I can answer any questions.

**Comment by willypuzzle** [comment]
@jdm Please assign me the issue then. 
Anyway, I tried but two branch are too different the https://github.com/servo/servo/pull/28302 is about 6500 commits behind `main`. Too many conflicts, I can try to copy the modifications by hand in the new branch having as root `main`.

**Comment by sagudev** [comment]
> Please assign me the issue then.

Done.

> Anyway, I tried but two branch are too different the [#28302](https://github.com/servo/servo/pull/28302) is about 6500 commits behind `main`. Too many conflicts, I can try to copy the modifications by hand in the new branch having as root `main`.

In that time we moved from compiler plugins to crown so yeah, it does make sense to start on more fresh base, but use PR as reference.

**Comment by willypuzzle** [comment]
@jdm @sagudev it's too difficult for me rebuild from the PR I don't understand and it's maybe unrelated to the current state of code. Maybe if you summarize the steps I should do I can build a PR from scratch.
Please, let me know.

**Comment by sagudev** [comment]
Actually, looking at this again, I think we should do it incrementally (or maybe in smaller scope by avoiding final step):

1. Impl empty drop for all DOM types (`#[dom_struct]` or maybe even `JSTraceable`) and allow custom drop impl (that would be provided as separate unsafe trait):
```rust
trait DOMDrop {
     /// SAFETY: This function must not use any jsmanaged field (as it can already be GCed)
     unsafe fn drop(&mut self);
}
```
2. Create crown lint to enforce no JSManaged field in struct (DroppableStruct) and replace all unsafe drops with this (we might need https://github.com/servo/servo/issues/32102 for this).

DISCLAIMER: This is just a sketch and I would probably need to think this through more.

**Comment by jdm** [comment]
@willypuzzle https://github.com/servo/servo/issues/26488#issuecomment-640237906 describes the work that needs to happen right now. One PR per interface where the existing Drop implementation is refined and minimized would be best to work on his incrementally.

**Comment by willypuzzle** [comment]
@jdm @sagudev I looked for a interface with Drop trait and I found no one. Could you suggest me a struct where I can begin from in order to implement https://github.com/servo/servo/issues/26488#issuecomment-640237906 ?

**Comment by sagudev** [comment]
> [@jdm](https://github.com/jdm) [@sagudev](https://github.com/sagudev) I looked for a interface with Drop trait and I found no one. Could you suggest me a struct where I can begin from in order to implement [#26488 (comment)](https://github.com/servo/servo/issues/26488#issuecomment-640237906) ?

From top of my head there is https://github.com/servo/servo/blob/56882a3d5bc4d461a6e253e4a4a1812cca9c1630/components/script/dom/webgpu/gpubuffer.rs#L185 (it's safe because it only access non-jsmanaged fields inside)

You can patch codegen to impl Drop for all concrete interface types and see the errors.

**Comment by jdm** [comment]
#27423 has some interfaces with Drop implementations.

**Comment by willypuzzle** [comment]
@jdm I have a question, should I remove the original Drop trait implementation? or should I leave both?

**Comment by jdm** [comment]
Yes, the goal is to remove the Drop implementation for structs with the `#[dom_struct]` annotation.

**Comment by willypuzzle** [comment]
I'm trying to undrop gpubuffer as @sagudev suggested. And how you can see [here](https://github.com/servo/servo/pull/36746) The result was that I moved all the members in the DroppableStruct, with unwanted side effects. @jdm What I need is a criteria to understand which fields need to be moved in the DrobbableStruct, related to the gpubuffer or in general. 

Thanks!

**Comment by jdm** [comment]
The only fields that need to be moved are the ones that are accessed inside of the Drop implementation.

**Comment by willypuzzle** [comment]
@jdm in order to discover dom_strucs with Drop trait I was used to add the Drop trait to the dom_struc macro. But now this way doesn't work anymore because many structs have Drop trait injected by config.
How can I solve?

**Comment by Gae24** [comment]
> @jdm in order to discover dom_strucs with Drop trait I was used to add the Drop trait to the dom_struc macro. But now this way doesn't work anymore because many structs have Drop trait injected by config.
> How can I solve?

Isn't enough to look for types that have `    allowDropImpl: True,` inside `Bindings.conf`?

**Comment by willypuzzle** [comment]
@sagudev this issue is not completed, my last PR fixed it only partially, please reopen it.

**Comment by sagudev** [comment]
I did some more thinking and we should implement empty drop on each (non-trivial) JSTraceable;
the only JSTraceables that can have drop impl are the ones that have #[no_trace] everywhere (trace impl is efectivelly nop).
Of course we would need an escape hatch (derive macro attribute `[derive()]`)

now in the light of https://github.com/servo/servo/pull/42180 we would also need drop impl on each dom_struct, so I propose we simply implement forbid custom drops of so we need to move all drops int dropable struct (even if they are unsafe at the present time).

**Comment by sagudev** [comment]
> I did some more thinking and we should implement empty drop on each (non-trivial) JSTraceable; the only JSTraceables that can have drop impl are the ones that have #[no_trace] everywhere (trace impl is efectivelly nop). Of course we would need an escape hatch (derive macro attribute `[derive()]`)

JSTraceables that are copy should also not have drop. Actually even empty drop has some side effects in rust, so ideally we would just pass negative bounds or smth. WIP: https://github.com/sagudev/servo/commit/9de7894500b920ff01f4b421220405d44ff0696b

**Comment by willypuzzle** [comment]
@sagudev practicaly what should I do?

**Comment by sagudev** [comment]
Creating droppable substructs as you did now.

**Comment by delan** [comment]
hi! i’m working on the monthly update, and i can see that a bunch of these changes have landed recently. what was the motivation for this effort? are Drop impls on DOM types unsound (or potentially unsound)? and at the same time, why is it ok for a #[dom_struct] type to *contain* another type with a Drop impl?

**Comment by jdm** [comment]
The unsoundness comes from a dom_struct type with a Drop implementation where it can interact with other dom_struct types (such as through Dom smart pointer members of the original type). Since dom_struct values don't have their Drop implementation called until the GC finalizer runs, sometimes these finalizers run in an order where the pointers are pointing to memory that has already been dropped. By pushing the Drop implementation into a different type we make it easier to restrict the data that can be accessed from the Drop implementation without affecting the normal use of the type being dropped.

**Comment by delan** [comment]
i see, thanks! so if we land a fix that makes a #[dom_struct] type contain a field of non-#[dom_struct] type that impls Drop, does that ensure we Drop the inner value *before* we run the GC finalizer for the outer value? and the GC finalizer in turn runs *before* we Drop the outer value, but that Drop will be empty and therefore harmless.

**Comment by jdm** [comment]
It doesn't change the ordering, but we can be more strict about keeping `Dom<T>` pointers out of types that have a Drop implementation, so when the values get dropped there's no way for the Drop implementation to access pointers that could be invalid.

**Comment by jdm** [comment]
The GC finaluzer is what causes the Drop to run for the dom_struct type.

**Comment by willypuzzle** [comment]
After the last PR only GPUBuffer and Promise struct are left.
Both are not simple cases, that require more investigation. 
I suggest to close this and open two related and specific issues.
@jdm @sagudev

**Comment by jdm** [comment]
Agreed! Filed #44639 and #44640. Thank you for all the PRs to get us to this point!
